network/wsNetwork.go

```diff
@@ -1144,6 +1162,40 @@ func (wn *WebsocketNetwork) ServeHTTP(response http.ResponseWriter, request *htt
 		challenge = wn.prioScheme.NewPrioChallenge()
 		responseHeader.Set(PriorityChallengeHeader, challenge)
 	}
+
+	var peerPublicKey crypto.PublicKey
+	var peerIDChallenge [32]byte
+	if shouldSupportIdentityChallenge(matchingVersion) {
+		idChalB64 := request.Header.Get(ProtocolConectionIdentityChallengeHeader)
+		if idChalB64 == "" {
+			wn.log.Warnf("ws connect no identification challenge")
+			return
+		}
+		idChal := IdentityChallengeFromB64(idChalB64)
+		// if the identification challenge response is not correctly signed, do not proceed
+		if err := idChal.verify(); err != nil {
+			wn.log.Warnf("ws connect failed identification challenge signature: err=%v", err.Error())
+			return
+		}
+		peerPublicKey = idChal.Key
+		// if we already have a verified peer using this key, do not proceed
+		wn.peersByIDLock.Lock()
+		defer wn.peersByIDLock.Unlock()
```

Comment:
in the case where this _doesn't_ return right away it holds the lock much longer than probably intended